### PR TITLE
Add allOf array && improve array/number mocker

### PR DESCRIPTION
**Pull request by sixaphone**

### CODE REVIEW

File: src/allOf.js [hunk 1]
```diff
@@ -27,14 +26,21 @@ export function allOfSample(into, children, options, spec) {
   if (res.type === 'object') {
     res.value = mergeDeep(res.value || {}, ...subSamples);
   } else {
-    // TODO: implement arrays
-    if (res.type === 'array' && !options.quiet) {
-      console.warn('OpenAPI Sampler: found allOf with "array" type. Result may be incorrect');
+    if (res.type === 'array' ) {
```
Comment (derevnjuk) [inline src/allOf.js]: Use `switch` or merge `if/else` declarations

File: src/utils.js [hunk 2]
```diff
@@ -48,3 +49,24 @@ export function mergeDeep(...objects) {
...
+    }
+
+    if (key === 'type') {
+      type = schema[key];
+    }
+
+    if (isObject(schema[key])) {
+      type = nestedTypeLookup(schema[key]);
+    }
+  }
```
Comment (derevnjuk) [inline src/utils.js]: Really? Well, obviously, you can write it easier.

File: src/utils.js [hunk 3]
```diff
@@ -48,3 +49,24 @@ export function mergeDeep(...objects) {
     return prev;
   }, Array.isArray(objects[objects.length - 1]) ? [] : {});
 }
+
+export function nestedTypeLookup(schema) {
```
Comment (derevnjuk) [inline src/utils.js]: What is it doing at all? OAS clearly defines available keywords.

File: src/allOf.js [hunk 4]
```diff
@@ -27,14 +26,21 @@ export function allOfSample(into, children, options, spec) {
...
+
+      if (!options.quiet) {
+        console.warn('OpenAPI Sampler: found allOf with "array" type. Result may be incorrect');
+      }
+
```
Comment (derevnjuk) [inline src/allOf.js:35]: You should check and throw an exception on schemas with different types here

File: src/allOf.js [hunk 5]
```diff
@@ -27,14 +26,21 @@ export function allOfSample(into, children, options, spec) {
...
+      const arraySchema = mergeDeep(...children);
+      if (!arraySchema.items) {
```
Comment (derevnjuk) [inline src/allOf.js]: `items` is required in `array`. See https://swagger.io/docs/specification/data-models/data-types/#array

File: src/samplers/array.js [hunk 6]
```diff
@@ -22,7 +46,17 @@ export function sampleArray(schema, options = {}, spec) {
...
+      throw new Error('Unable to generate so many unique values for boolean.');
+    }
+
+    let {value: sample} = traverse(itemSchema, options, spec);
+    if (schema.uniqueItems) {
+      while (hasItem(res, sample, itemSchema.type)) {
+         sample = traverse(itemSchema, options, spec).value;
+      }
+    }
+
```
Comment (derevnjuk) [inline src/samplers/array.js]: We don't need to support `uniqueItems`. It's unhelpful in mock context

File: test/integration.spec.js [hunk 7]
```diff
@@ -234,6 +234,50 @@ describe('Integration', function() {
       expect(Array.isArray(result.arr)).to.equal(true);
     });
 
+    it('allOf array of at least two numbers', function() {
```
Comment (derevnjuk) [inline test/integration.spec.js]: Bring test titles into line with others.

**Review by derevnjuk — CHANGES_REQUESTED** [review]

File: src/allOf.js (hunk 4 above)
Comment (sixaphone) [reply src/allOf.js:35]: is that not handled by the loop above, the `subSamples`.
Or do you mean array types

Comment (derevnjuk) [reply src/allOf.js:35]: We've already discussed it with us. The types can't be merged, if they are logical incompatible. It mean `string` can't be combined with `integer` and etc.

Comment (derevnjuk) [reply src/allOf.js:35]: @sixaphone Seems it's my fault. We've already handled that case.
https://github.com/NeuraLegion/openapi-sampler/blob/ec9fae76ddd78e4931d347993cd4ccacbad0d114/src/allOf.js#L14
Following schema is allowed also. It mean `array` can contain both string and number items.

```json
{
   "allOf": [
      {
         "type": "array",
         "items": {
            "type": "number"
         }
      },
      {
         "items": {
            "type": "string"
         },
         "minItems": 2,
         "maxItems": 3
      }
   ]
}
```

Comment (sixaphone) [reply src/allOf.js:35]: :sweat_smile:  its okay i still love you, no homo

**Review by derevnjuk — APPROVED** [review]

**Pull request merged**